pkg/operator/controller/instance_controller_test.go

```diff
@@ -0,0 +1,319 @@
+	testEnv := &envtest.Environment{
+		CRDDirectoryPaths:     []string{filepath.Join("..", "..", "..", "config", "operator", "crds")},
+		ErrorIfCRDPathMissing: true,
+		BinaryAssetsDirectory: filepath.Join("..", "..", "..", "bin", "k8s",
+			fmt.Sprintf("1.28.3-%s-%s", runtime.GOOS, runtime.GOARCH)),
+	}
+
+	var err error
+	cfg, err := testEnv.Start()
+	if err != nil {
+		fmt.Printf("Failed to start test environment: %v\n", err)
+		cancel()
+		return
+	}
+
+	err = clusterlink.AddToScheme(scheme.Scheme)
+	if err != nil {
+		fmt.Printf("Failed to add API to scheme: %v\n", err)
+		cancel()
+		return
+	}
+
+	k8sClient, err = client.New(cfg, client.Options{Scheme: scheme.Scheme})
+	if err != nil {
+		fmt.Printf("Failed to create Kubernetes client: %v\n", err)
+		cancel()
+		return
+	}
+
+	k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
+		Scheme: scheme.Scheme,
+	})
+	if err != nil {
+		fmt.Printf("Failed to create controller manager: %v\n", err)
+		cancel()
+		return
+	}
+
+	err = (&controller.InstanceReconciler{
+		Client:        k8sManager.GetClient(),
+		Scheme:        k8sManager.GetScheme(),
+		Logger:        logrus.WithField("component", "reconciler"),
+		InstancesMeta: make(map[string]string),
+	}).SetupWithManager(k8sManager)
+	if err != nil {
+		fmt.Printf("Failed to set up reconciler: %v\n", err)
+		cancel()
+		return
+	}
+
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				fmt.Printf("Recovered from panic: %v\n", r)
+			}
+		}()
+		err := k8sManager.Start(ctx)
+		if err != nil {
+			fmt.Printf("Failed to run controller manager: %v\n", err)
+		}
+	}()
+
+	code := m.Run()
+
+	// Cancelling the context
+	cancel()
+	// Tearing down the test environment
+	err = testEnv.Stop()
+	if err != nil {
+		fmt.Printf("Failed to stop test environment: %v\n", err)
+	}
+
+	// Exiting with the status code from the test run
+	os.Exit(code)
+}
+
+func TestClusterLinkController(t *testing.T) {
+	RegisterTestingT(t) // Setup gomega
+	// Define utility constants for object names and testing timeouts/durations and intervals.
+
+	cl := clusterlink.Instance{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "batch.tutorial.kubebuilder.io/v1",
+			Kind:       "Clusterlink",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ClusterLinkName,
+			Namespace: controller.OperatorNameSpace,
+		},
+	}
+
+	cpID := types.NamespacedName{Name: controller.ControlPlaneName, Namespace: controller.InstanceNamespace}
+	cpResource := []client.Object{
+		&appsv1.Deployment{}, &corev1.Service{}, &corev1.PersistentVolumeClaim{},
+		&rbacv1.ClusterRole{}, &rbacv1.ClusterRoleBinding{},
+	}
+	dpID := types.NamespacedName{Name: controller.DataPlaneName, Namespace: controller.InstanceNamespace}
+	dpResource := []client.Object{&appsv1.Deployment{}, &corev1.Service{}}
+	ingressID := types.NamespacedName{Name: controller.IngressName, Namespace: controller.InstanceNamespace}
+
+	t.Run("Create ClusterLink deployment", func(t *testing.T) {
```

Comment:
Is the gomega style testing required? It differs from everything else we have in the code base.

Reply:
I need it just for the "Eventually" function that uses deployed units when we create them, for all other cases I can use assert, but I thought to keep it in the same style for this specific test.

Reply:
Please use a consistent style as other tests in the repo, not using Gomega style.
If you need a specific function, such as Eventually, maybe it's available in testify already or can be implemented locally?